### ISSUE TITLE
Enhance/New Lint: `option_and_then_some` for the `Result` equivalent

### ISSUE THREAD

**Issue by flip1995** [rust-lang/rust-clippy#5492]
It does **not** (yet) report anything for the `Result` equivalent:

```rust
fn main() {
    let a = Err::<(), _>("error");
    let b = a.or_else(|_| Err("another error"));
    println!("{:?}", b);
}
```

_Originally posted by @shepmaster in https://github.com/rust-lang/rust-clippy/issues/1459#issuecomment-616199145_

**Comment by shepmaster** [comment]
Maybe also for `Result::and_then` (as [mentioned by the original suggester](https://github.com/rust-lang/rust-clippy/issues/1459#issuecomment-414766838), @drystone):

```rust
fn main() {
    let a = Ok::<_, ()>(42);
    let b = a.and_then(|_| Ok(22));
    println!("{:?}", b);
}
```

**Comment by shepmaster** [comment]
In the first case, I'd expect suggestion to use `map_err`, the second would be `map`.

**Comment by alex-700** [comment]
Hi! If you don't mind, I'll add this to the weekends. Is it better to introduce new lint for it or append to the current (and rename it)?

**Comment by flip1995** [comment]
You can introduce a new lint for this `result_and_then_ok`. Maybe also `result_or_else_err`, but I'm not sure if we should split those up that far.

**Comment by alex-700** [comment]
I decided to significantly improve the lint `option_and_then_some` before generalization to `Result` cases. More specifically I want to get this lint not only on
```rust
let _ = x.and_then(|o| Some(o + 1));
```
but on 
```rust
let _ = x.and_then(|o| {
    do_smth(o);
    Some(o + 1)
})
```
and even on 
```rust
let _ = opt().and_then(|s| {
    if s.len() < 2 {
        return Some(20);
    } else if s.len() > 20 {
        return Some(s.len());
    }
    Some(30)
});
```

And I did it. And waste a lot of time to understand that multisuggestion doesn't work (https://github.com/rust-lang/rust/issues/53934).
And now I have a question: is it ok to drop rust-fix ui tests (comment with FIXME after 53934)? Or autofix functionality (only for `|s| Some(s)`) is more important than possibility to find more interesting cases?

Also I want to unite `Option` and `Result` lints into one. What do you think about `lazy_logical_instead_of_map` name? Or even more "monadic": `bind_instead_of_map`?

**Comment by flip1995** [comment]
>  is it ok to drop rust-fix ui tests (comment with FIXME after 53934)?

Maybe use `db.span_suggestion` for the simple case and only use `multipart_sugg` for the other case. But I guess you already did this. That way the simple case is still auto fixable. For testing: just split up the tests in two files: <lint_name>.rs and <lint_name>_multipart.rs, where only on the first file rustfix is run.

As for the lint name: I like the second one better, so I suggest you go with that for now, and we can revisit the naming just before merging the PR (renaming should not be that much work anyway).
